### ISSUE TITLE
[Paper Note] Unified-IO 2: Scaling Autoregressive Multimodal Models with Vision, Language, Audio, and Action, Jiasen Lu+, CVPR'24, 2023.12

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#1202]
# URL
- https://arxiv.org/abs/2312.17172
# Authors
  - Jiasen Lu
  - Christopher Clark
  - Sangho Lee
  - Zichen Zhang
  - Savya Khosla
  - Ryan Marten
  - Derek Hoiem
  - Aniruddha Kembhavi
# Abstract
  - We present Unified-IO 2, the first autoregressive multimodal model that is capable of understanding and generating image, text, audio, and action. To unify different modalities, we tokenize inputs and outputs -- images, text, audio, action, bounding boxes, etc., into a shared semantic space and then process them with a single encoder-decoder transformer model. Since training with such diverse modalities is challenging, we propose various architectural improvements to stabilize model training. We train our model from scratch on a large multimodal pre-training corpus from diverse sources with a multimodal mixture of denoisers objective. To learn an expansive set of skills, such as following multimodal instructions, we construct and finetune on an ensemble of 120 datasets with prompts and augmentations. With a single unified model, Unified-IO 2 achieves state-of-the-art performance on the GRIT benchmark and strong results in more than 35 benchmarks, including image generation and understanding, natural language understanding, video and audio understanding, and robotic manipulation. We release all our models to the research community.
# Translation (by gpt-4o-mini)
- 私たちは、画像、テキスト、音声、アクションを理解し生成できる初の自己回帰型マルチモーダルモデル「Unified-IO 2」を紹介します。異なるモダリティを統一するために、画像、テキスト、音声、アクション、バウンディングボックスなどの入力と出力を共有された意味空間にトークン化し、1つのエンコーダ-デコーダ型トランスフォーマーモデルで処理します。このような多様なモダリティを用いてのトレーニングは難易度が高いため、モデルのトレーニングを安定させるためのさまざまなアーキテクチャの改善を提案します。私たちは、多様なソースからの大規模なマルチモーダル事前学習コーパスでモデルをゼロからトレーニングし、マルチモーダルな混合除去器の目的で訓練します。マルチモーダルな指示に従うなどの広範なスキルセットを学ぶために、プロンプトと拡張を用いて120のデータセットのアンサンブルで構成し微調整します。Unified-IO 2は、単一の統一モデルでGRITベンチマークにおいて最先端のパフォーマンスを達成し、画像生成と理解、自然言語理解、映像と音声理解、ロボット操作を含む35以上のベンチマークで強力な結果を示します。私たちは、すべてのモデルを研究コミュニティに公開します。
# Summary (by gpt-4o-mini)
- 初の自己回帰型マルチモーダルモデル「Unified-IO 2」を提案し、画像、テキスト、音声、アクションを統一した意味空間で処理。トレーニングの安定化のためにアーキテクチャを改善し、120のデータセットで微調整を行い、GRITベンチマークで最先端のパフォーマンスを達成。35以上のベンチマークにおいて強力な結果を示し、すべてのモデルを公開。

**Comment by AkihikoWatanabe** [comment]
画像、テキスト、音声、アクションを理解できる初めてのautoregressive model。AllenAI

**Comment by AkihikoWatanabe** [comment]
モデルのアーキテクチャ図
![image](https://github.com/user-attachments/assets/4282ffb0-18f1-40c9-b6d7-f004d03b8382)

マルチモーダルに拡張したことで、訓練が非常に不安定になったため、アーキテクチャ上でいくつかの工夫を加えている:

- 2D Rotary Embedding
  - Positional EncodingとしてRoPEを採用
  - 画像のような2次元データのモダリティの場合はRoPEを2次元に拡張する。具体的には、位置(i, j)のトークンについては、Q, Kのembeddingを半分に分割して、それぞれに対して独立にi, jのRoPE Embeddingを適用することでi, j双方の情報を組み込む。
- QK Normalization
  - image, audioのモダリティを組み込むことでMHAのlogitsが非常に大きくなりatteetion weightが0/1の極端な値をとるようになり訓練の不安定さにつながった。このため、dot product attentionを適用する前にLayerNormを組み込んだ。
- Scaled Cosine Attention
  - Image Historyモダリティにおいて固定長のEmbeddingを得るためにPerceiver Resamplerを扱ったているが、こちらも上記と同様にAttentionのlogitsが極端に大きくなったため、cosine類似度をベースとしたScaled Cosine Attention #2259 を利用することで、大幅に訓練の安定性が改善された。
- その他
  - attention logitsにはfp32を適用
  - 事前学習されたViTとASTを同時に更新すると不安定につながったため、事前学習の段階ではfreezeし、instruction tuningの最後にfinetuningを実施

![image](https://github.com/user-attachments/assets/74c8fa3a-8fb5-4785-8dd3-6a8cf3c7cfeb)

**Comment by AkihikoWatanabe** [comment]
目的関数としては、Mixture of Denoisers (#1424)に着想を得て、Multimodal Mixture of Denoisersを提案。MoDでは、
- \[R\]: 通常のspan corruption (1--5 token程度のspanをmaskする)
- \[S\]: causal language modeling (inputを2つのサブシーケンスに分割し、前方から後方を予測する。前方部分はBi-directionalでも可)
- \[X\]: extreme span corruption (12>=token程度のspanをmaskする)

の3種類が提案されており、モダリティごとにこれらを使い分ける:
- text modality: UL2 (#1424)を踏襲
- image, audioがtargetの場合: 2つの類似したパラダイムを定義し利用
  - \[R\]: patchをランダムにx%マスクしre-constructする
  - \[S\]: inputのtargetとは異なるモダリティのみの情報から、targetモダリティを生成する

訓練時には prefixとしてmodality token \[Text\], \[Image\], \[Audio\] とparadigm token \[R\], \[S\], \[X\] をタスクを指示するトークンとして利用している。

**Comment by AkihikoWatanabe** [comment]
また、image, audioのマスク部分のdenoisingをautoregressive modelで実施する際には普通にやるとdecoder側でリークが発生する(a)。これを防ぐには、Encoder側でマスクされているトークンを、Decoder側でteacher-forcingする際にの全てマスクする方法(b)があるが、この場合、生成タスクとdenoisingタスクが相互に干渉してしまいうまく学習できなくなってしまう（生成タスクでは通常Decoderのinputとして[mask]が入力され次トークンを生成する、といったことは起きえないが、愚直に(b)をやるとそうなってしまう）。ので、(c)に示したように、マスクされているトークンをinputとして生成しなければならない時だけ、マスクを解除してdecoder側にinputする、という方法 (Dynamic Masking) でこの問題に対処している。
<img width="597" height="394" alt="Image" src="https://github.com/user-attachments/assets/0dba8d5d-0c93-4c56-852b-fce9869428e7" />
